include/swift/AST/Stmt.h

```diff
@@ -1110,15 +1110,19 @@ class SwitchStmt final : public LabeledStmt,
   friend TrailingObjects;
 
   SourceLoc SwitchLoc, LBraceLoc, RBraceLoc;
+  /// The location of the last token in the 'switch' statement. For valid
+  /// 'switch' statements this is the same as \c RBraceLoc. If the '}' is
+  /// missing this points to the last token before the '}' was expected.
+  SourceLoc EndLoc;
```

Comment:
Instead of introducing `EndLoc`, can we pass `PreviousLoc` to `LBraceLoc` and `RBraceLoc`  in Parser?
This way, we can know "invalid" switch statements by checking `LBraceLoc == RBraceLoc`.

Reply:
We could… I also thought about just abusing `RBraceLoc` to store the end location like we do for `BraceStmt`. I think after the parser, no-one really cares about the `LBraceLoc` and `RBraceLoc`. But I didn’t really like that approach for `BraceStmt` already and I don’t want to make the situation worse. 

I don’t know if there will ever be a use case for it but we wouldn’t be able to check whether there was a `{` but no `}` with your approach. And we would have `LBraceLoc` be after the end loc of the last case if just the `}` is missing, which isn’t nice either.

As an alternative, what do you think about renaming `RBraceLoc` to `EndLoc` and adding a boolean flag `HasRBrace`. If it is `true`, `EndLoc` represents the `RBraceLoc` and if it is `false` there was no `}`.

Reply:
I agree abusing `RBraceLoc` is not nice.
It's OK. As long as you intentionally do this, I support it :)